src/platform/with_eagl/native_gl_context.rs

```diff
@@ -0,0 +1,128 @@
+use platform::NativeGLContextMethods;
+use GLVersion;
+use gleam::gl;
+use objc::runtime::{BOOL, NO};
+use objc::runtime::{Class, Object};
+use libloading as lib;
+use std::ops::Deref;
+
+type EAGLContext = *mut Object;
+type EAGLSharegroup = *mut Object;
+
+lazy_static! {
+    static ref EAGLCONTEXT_CLASS: &'static Class  = {
+        Class::get("EAGLContext").expect("EAGLContext class not found")
+    };
+
+    static ref OPENGLES_FRAMEWORK: Option<lib::Library>  = {
+        lib::Library::new("/System/Library/Frameworks/OpenGLES.framework/OpenGLES").ok()
+    };
+}
+
+pub struct NativeGLContext(EAGLContext);
+
+impl Drop for NativeGLContext {
+    fn drop(&mut self) {
+        unsafe {
+            msg_send![self.0, release];
+        }
+    }
+}
+
+unsafe impl Send for NativeGLContext {}
+
+pub struct NativeGLContextHandle(EAGLContext);
+
+unsafe impl Send for NativeGLContextHandle {}
+
+impl NativeGLContextMethods for NativeGLContext {
+    type Handle = NativeGLContextHandle;
+
+    fn get_proc_address(addr: &str) -> *const () {
+        if let Some(ref lib) = *OPENGLES_FRAMEWORK {
+            unsafe {
+                let symbol: Result<lib::Symbol<unsafe extern fn()>, _> = lib.get(addr.as_bytes());
+                if let Ok(symbol) = symbol {
+                    return *symbol.deref() as *const ();
+                }
+            }
+        }
+        0 as *const _
+    }
+
+    fn create_shared(with: Option<&Self::Handle>,
+                     _api_type: &gl::GlType,
+                     api_version: GLVersion) -> Result<NativeGLContext, &'static str> {
+        let context: EAGLContext = unsafe {
+            let context: EAGLContext = msg_send![*EAGLCONTEXT_CLASS, alloc];
+            let version = api_version.major_version() as u32;
+            if let Some(with) = with {
+                let sharegroup: EAGLSharegroup = msg_send![with.0, sharegroup];
+                msg_send![context, initWithAPI: version sharegroup: sharegroup]
+            } else {
+                msg_send![context, initWithAPI: version]
+            }
+        };
+
+        if context.is_null() {
+            Err("[EAGLContext initWithAPI] failed")
+        } else {
+            Ok(NativeGLContext(context))
+        }
+    }
+
+    fn is_current(&self) -> bool {
+        let context: EAGLContext = unsafe { 
+            msg_send![*EAGLCONTEXT_CLASS, currentContext] 
+        };
+        context == self.0
+    }
+
+    fn current() -> Option<Self> {
+        let context: EAGLContext = unsafe {
+            msg_send![*EAGLCONTEXT_CLASS, currentContext]
+        };
+        if context.is_null() {
+            None
+        } else {
+            Some(NativeGLContext(context))
```

Comment:
So, this is creating a `NativeGLContext`, which will `release` it when it gets dropped.

However, the call above, and below, don't seem to `addref` it in any way.

So either calling this function will eventually lead to a double-free of the context, or the other two are leaking the context.

Reply:
Yes, that would lead to double-frees or leaked references. I have retained all EAGLContext instances in NativeGLContext and NativeGLContextHandle structs now.